File: backend/app/schemas/vector_index.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.vector_index import IndexVersionStatus, VectorIndexStatus


class VectorIndexRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    embedding_version_id: uuid.UUID
    document_id: uuid.UUID
    provider: str
    index_type: str
    namespace: str
    dimensions: int
    version: int
    status: VectorIndexStatus
    status_message: str | None
    vector_count: int
    build_duration_ms: int | None
    created_at: datetime
    updated_at: datetime


class IndexVersionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    vector_index_id: uuid.UUID
    version: int
    vector_count: int
    status: IndexVersionStatus
    status_message: str | None
    build_duration_ms: int | None


class CreateVectorIndexRequest(BaseModel):
    provider: str = Field(..., min_length=1, max_length=20)
    index_type: str = Field(default="hnsw", min_length=1, max_length=20)

File: frontend/src/types/vector-index.ts
export type VectorIndexProviderName = 'pgvector' | 'qdrant' | 'chroma' | 'pinecone';

export const VECTOR_INDEX_PROVIDERS: { value: VectorIndexProviderName; label: string }[] = [
  { value: 'pgvector', label: 'PgVector (default)' },
  { value: 'qdrant', label: 'Qdrant' },
  { value: 'chroma', label: 'Chroma' },
  { value: 'pinecone', label: 'Pinecone' },
];

export type VectorIndexType = 'hnsw' | 'ivf_flat' | 'flat' | 'pq';

export const VECTOR_INDEX_TYPES: { value: VectorIndexType; label: string }[] = [
  { value: 'hnsw', label: 'HNSW' },
  { value: 'ivf_flat', label: 'IVF Flat' },
  { value: 'flat', label: 'Flat (exact)' },
  { value: 'pq', label: 'Product quantization' },
];

export type VectorIndexStatus = 'pending' | 'building' | 'ready' | 'failed';

export interface VectorIndex {
  id: string;
  embedding_version_id: string;
  document_id: string;
  provider: string;
  index_type: string;
  namespace: string;
  dimensions: number;
  version: number;
  status: VectorIndexStatus;
  status_message: string | null;
  vector_count: number;
  build_duration_ms: number | null;
  created_at: string;
  updated_at: string;
}
